src/lib/sql/sql_pipeline_statement.cpp

```diff
@@ -97,45 +102,119 @@ const std::shared_ptr<AbstractLQPNode>& SQLPipelineStatement::get_unoptimized_lo
   return _unoptimized_logical_plan;
 }
 
+void SQLPipelineStatement::split_expression(std::shared_ptr<AbstractExpression> &expression,
+                                            std::vector<std::shared_ptr<AbstractExpression>>& values,
+                                            ParameterID &next_parameter_id) {
+  if (expression->type == ExpressionType::Value) {
+    if (expression->replaced_by) {
+      const auto valexp = std::dynamic_pointer_cast<ValueExpression>(expression);
+      expression = expression->replaced_by;
+    } else {
+      const auto valexp = std::dynamic_pointer_cast<ValueExpression>(expression);
+      if (valexp->data_type() != DataType::Null) {
+        assert(expression->arguments.empty());
+        values.push_back(expression);
+        auto new_expression =
+            std::make_shared<TypedPlaceholderExpression>(next_parameter_id, expression->data_type());
+        expression->replaced_by = new_expression;
+        expression = new_expression;
+        next_parameter_id++;
+      }
+    }
+  }
+}
+
+const std::shared_ptr<AbstractLQPNode>& SQLPipelineStatement::get_split_unoptimized_logical_plan(
+    std::vector<std::shared_ptr<AbstractExpression>>& values) {
+  auto& unoptimized_lqp = get_unoptimized_logical_plan();
+
+  ParameterID next_parameter_id(0);
+
+  if (_translation_info.cacheable) {
+    visit_lqp(unoptimized_lqp, [&values, &next_parameter_id](const auto& node) {
+      if (node) {
+        for (auto& root_expression : node->node_expressions) {
+          visit_expression(root_expression, [&values, &next_parameter_id](auto& expression) {
+            split_expression(expression, values, next_parameter_id);
+            return ExpressionVisitation::VisitArguments;
+          });
+        }
+      }
+      return LQPVisitation::VisitInputs;
+    });
+  }
+
+  return unoptimized_lqp;
+}
+
 const std::shared_ptr<AbstractLQPNode>& SQLPipelineStatement::get_optimized_logical_plan() {
   if (_optimized_logical_plan) {
     return _optimized_logical_plan;
   }
 
+  const auto started_preoptimization_cache = std::chrono::high_resolution_clock::now();
+
+  std::vector<std::shared_ptr<AbstractExpression>> values;
+  const auto unoptimized_lqp = get_split_unoptimized_logical_plan(values);
+
   // Handle logical query plan if statement has been cached
   if (lqp_cache) {
-    if (const auto cached_plan = lqp_cache->try_get(_sql_string)) {
-      const auto plan = *cached_plan;
+    if (const auto cached_plan = *(lqp_cache->try_get(unoptimized_lqp))) {
```

Comment:
Why do we need the `*` here?

Reply:
Yes, we think this is necessary to unwrap the optional we receive from `try_get()`.